Add loading step concept to Tensor Mechanics

recuero opened the pull request:
## Reason
Users may want to organize loading steps such that they can use their own material routines and perform boundary conditions/constraints enabling/disabling activities.

## Design
Create tailored controls. Inform user plugins of step number.

## Impact
More user-friendly interface --at least to some users.

recuero committed:
Test that can be used in final implementation (#19249)

recuero committed:
Add loading step (STEP) concept to Tensor Mechanics (#19249)

This commit does a few things:

 - Add a time period control (StepPeriod) that's driven by loading steps
 - Add a general user object that generates steps based on simulation times
 - Make AbaqusUMATStress and AbaqusUExternalDB work with actual steps and pass
   the step number to user plugins.
 - Minor refactor of TimePeriod control in the framework

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1367075/) on 776babf wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23509/site)

This comment will be updated on new commits.

dschwen commented on modules/tensor_mechanics/src/materials/abaqus/AbaqusUMATStress.C:
```diff
@@ -44,6 +45,8 @@ AbaqusUMATStress::validParams()
   params.addParam<MooseEnum>("decomposition_method",
                              ComputeFiniteStrain::decompositionType(),
                              "Method to calculate the strain kinematics.");
+  params.addParam<UserObjectName>(
```
It would be nice if that object could be obtained automatically

recuero replied:
Automatically by casting? If so, it may be risky as there could be multiple `StepUserObject`s for whatever reason.

dschwen replied:
Yeah, we discussed that, and I'm of the opinion that this should trigger an error. I don;t like that this creates another pitfall for the users. They add the steps but forget to specify the parameter to use the steps.

recuero replied:
We certainly didn't discuss the possible cases that appear as we enforce this behavior. I think doing what you suggest is fine, but comes at the expense of limiting the the possibilities of using steps.

dschwen replied:
We could have it both ways by searching for a userobjec if it isn't supplied explicitly. The same pattern is used in the ray tracing module, where the _RayTracingStudy_ is an optional parameter to RayKernels and RayBCs. If that parameter is not supplied the RayTracingStudy object is automatically detected, but it has to be unique.

I did mention that having  multiple `StepUserObject`s should not be permitted. But to quote George Bernard Shaw: "The single biggest problem in communication is the illusion that it has taken place".

dschwen commented on modules/tensor_mechanics/src/userobjects/AbaqusUExternalDB.C:
```diff
@@ -37,8 +37,11 @@ AbaqusUExternalDB::AbaqusUExternalDB(const InputParameters & parameters)
+    _aqSTEP(0),
+    _current_execute_on_flag(_fe_problem.getCurrentExecuteOnFlag()),
+    _step_user_object(isParamSetByUser("step_user_object")
+                          ? &getUserObject<StepUserObject>("step_user_object")
```
Again, automatic would be my preference. And this might have to go into inotialSetup

dschwen commented on modules/tensor_mechanics/include/userobjects/StepUserObject.h:
```diff
@@ -0,0 +1,41 @@
+  StepUserObject(const InputParameters & parameters);
+
+  // Get time
+  Real getInitTime(const unsigned int & step) const;
```
getStartTime maybe?

recuero committed:
Fix tests/docs.

Address reviewer comment.

dschwen commented on framework/src/controls/TimePeriodBase.C:
```diff
@@ -0,0 +1,68 @@
+bool
+TimePeriodBase::conditionMet(const unsigned int & /*i*/)
+{
+  mooseError("Internal error: TimePeriodBase::initialSetup() should not be called");
```
fix error message

dschwen commented on modules/tensor_mechanics/doc/content/source/controls/StepPeriod.md:
```diff
@@ -0,0 +1,24 @@
+i.e. steps, over which the boundary conditions and contraints are not disabled/enabled.
+A step number is thus a required input to this control object. At the beginning of this step number,
+boundary conditions and constraints are enabled or disabled, as directed by
+[StepUserObject](/StepUserObject.md). Only one step is allowed per `StepPeriod` object.
```
> Only one step is allowed per `StepPeriod` object.

This is something we could easily change in the future and would be a useful functionality.

dschwen commented on modules/tensor_mechanics/src/userobjects/StepUserObject.C:
```diff
@@ -0,0 +1,96 @@
+
+  params.addRequiredParam<std::vector<Real>>(
+      "times",
+      "The times at which the objects are to be enabled/disabled at the beginning of the steps.");
```
These are the start times for each step?

recuero replied:
I am making the guess that you are suggesting to have a more elaborate name, such as `step_start_times`?

dschwen commented on modules/tensor_mechanics/src/userobjects/StepUserObject.C:
```diff
@@ -0,0 +1,96 @@
+  params.addRequiredParam<std::vector<Real>>(
+      "times",
+      "The times at which the objects are to be enabled/disabled at the beginning of the steps.");
+  params.addRequiredParam<unsigned int>(
+      "number_steps",
+      "The number of loading steps defined for the simulation. The input 'times' "
+      "must have the same size as the number of steps.");
```
Any reason for adding this parameter? Why not just deduce it from the number of supplies start times?

recuero replied:
I can do that. It's not clear to me that it is better to the user. I will modify the times definition to make it more verbose. Hopefully users read it.

dschwen commented on modules/tensor_mechanics/src/userobjects/StepUserObject.C:
```diff
@@ -0,0 +1,96 @@
+Real
+StepUserObject::getStartTime(const unsigned int & step) const
+{
+  if (_times.size() < step)
```
```suggestion
  if (_times.size() <= step)
```

dschwen commented on modules/tensor_mechanics/src/userobjects/StepUserObject.C:
```diff
@@ -0,0 +1,96 @@
+
+  for (const auto i : index_range(_times))
+  {
+    if (i == _times.size())
+      break;
```
this never happens

recuero replied:
oh yeah, I had a different iteration strategy

recuero replied:
Same thing here. A check is still required.

dschwen commented on modules/tensor_mechanics/src/userobjects/StepUserObject.C:
```diff
@@ -0,0 +1,96 @@
+      return which_step;
+  }
+
+  return which_step;
```
the `_times` vector is sorted, so all you need to check is if `time >= _times[i]` and just immediately `return i;` if that's the case

recuero replied:
Sure. Not that we are optimizing anything meaningful.

recuero replied:
Problem is immediately returning will give you the lowest step number. For example, 3 steps, 0 1 2. if time = 2.2, 2.2 > 0 so period is 0. So I basically kept the original lines.

dschwen replied:
Right, it should have been "loop until _times[i] > time then return i-1" and return the last step number if there is not start time > time

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1368513/) on a9a2848 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>db9e5c</th>
      <th colspan=4>#23509 a9a284</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.71%</td>
      <td>84.71%</td>
      <td>-0.00%</td>
      <td>86.21%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>83154</td>
      <td>83164</td>
      <td>+10</td>
      <td>25</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15010</td>
      <td>15014</td>
      <td>+4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23509/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23509/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>db9e5c</th>
      <th colspan=4>#23509 a9a284</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.25%</td>
      <td>89.26%</td>
      <td>+0.01%</td>
      <td>93.02%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26410</td>
      <td>26486</td>
      <td>+76</td>
      <td>80</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3180</td>
      <td>3186</td>
      <td>+6</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23509/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23509/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23509/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 86.21% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Make user object optional input to user plugins

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1369451/) on 9740e0d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23509/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23509/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format db9e5cb19febcf402cba7a3c5994b4ac36d668dd`

dschwen commented on modules/tensor_mechanics/src/materials/abaqus/AbaqusUMATStress.C:
```diff
@@ -131,6 +134,36 @@ AbaqusUMATStress::initialSetup()
+  if (!isParamSetByUser("step_user_object"))
+  {
+    std::vector<const UserObject *> uos;
+    _fe_problem.theWarehouse().query().condition<AttribSystem>("UserObject").queryInto(uos);
```
Do a `queryIntoUnsorted` here to avoid an unnecessary dependency resolution. You might think it's not an "important" optimization, but your call will enforce a stricter detection of cyclical dependencies than what MOOSE currently does (only within execute on groups). I observed exactly that in a current PR where an external app suddenly failed with a cyclic dependency error.

recuero committed:
Add features

 - StepPeriod also automatically detects StepUserObjects and errors out if it detects more than one
 - User can choose to input step durations or step number and total time interval (in addition to start times)
 - Synchronization of solution is true by default for the StepPeriod control

recuero committed:
Add warehouse functionality to common interface base class

dschwen commented on framework/src/controls/TimePeriodBase.C:
```diff
@@ -0,0 +1,68 @@
+      mooseError("The start time(s) must be less than the end time(s).");
+}
+
+void
+TimePeriodBase::initialSetup()
+{
+  mooseError("Internal error: TimePeriodBase::initialSetup() should not be called");
+}
```
Can you remove the method entirely?

dschwen replied:
Removing this and the one below would bump your coverage to >>90%

dschwen commented on framework/src/controls/TimePeriodBase.C:
```diff
@@ -0,0 +1,68 @@
+}
+
+bool
+TimePeriodBase::conditionMet(const unsigned int & /*i*/)
+{
+  mooseError("Internal error: TimePeriodBase::conditionMet() should not be called");
+
+  return false;
+}
```
Same here. Could this be a pure virtual method?

dschwen replied:
I think it could

recuero replied:
Virtual now.

dschwen reviewed (dismissed):
Take my comments as suggestions. In the spirit of today's discussion this is good enough to get merged.

recuero commented:
Too late. That's going stale.

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1379330/) on 0b1127a : invalidated by @recuero
